### ISSUE TITLE
Add parameter-penalty hook to GMM (estimator-side regularization, not artificial moments)

### ISSUE THREAD

**Issue by ligon** [ligon/ManifoldGMM#19]
## Scope split: MR1 vs. deferred

Adding this section after design discussion to make the implementation scope unambiguous.  Original body below is unchanged.

### What MR1 ships

The natural first PR cuts the feature roughly in half along the inference axis.  Lands together:

- `GMM(..., penalty=...)`: accepts either a bare `Callable[[Any], float]` (autodiff-only, JAX-friendly) or a `PenaltyStrategy` Protocol with `value(theta)`, optional `gradient(theta)`, optional `hessian(theta)`.  The Protocol mirrors `WeightingStrategy` and lets `pymanopt.function.numpy` users supply analytic derivatives instead of paying finite-difference cost on every outer iter.  Callable form is sugar over the Protocol.
- `GMMResult.criterion_value` (existing): stays the optimizer's stopping cost, penalty included.
- `GMMResult.data_criterion_value` (new -- renamed from the original `unpenalized_criterion_value` proposal): the data-only `g_bar' W g_bar` at `θ̂_pen`.  Docstring states explicitly that this is **not** asymptotically χ²-distributed under penalization (the unpenalized FOC is not solved at `θ̂_pen`, so the leading term in the expansion picks up an `O(λ)` bias).  Useful as a fit diagnostic at the chosen point, not as an overidentification test.
- `compute_hessian_cond(data_only: bool = False)`: existing return value (`cond(D'WD + ∇²p)`) becomes the default; `data_only=True` returns the original `cond(D'WD)`.  Bit-identical to current behaviour when `penalty is None`.
- `GMMResult.tangent_covariance`: information matrix becomes `D'WD + ∇²p(θ̂_pen)`; outer-product-of-score remains data-only (penalty contributes no variance).  Docstring states the sandwich is valid for inference **about `θ̂_pen`** -- which is an asymptotically biased estimator of `θ_0`.  Frequentist coverage of `θ_0` requires bias-aware methods that are out of scope here.
- `GMMResult.k_statistic`: raises `NotImplementedError` pointing at #21 when called on a `GMMResult` produced with `penalty is not None`.  Unpenalized path unchanged.
- `weighting_info`/`optimizer_report`: extended to carry penalty configuration so pickled results are self-describing.
- Tests 1, 2, 3, 4, 6 from the original test list below.

### What's deferred to follow-ups

- **K-statistic under penalty** -- tracked in **#21**.  The derivation is non-trivial (penalized score, mixed identification/over-identification under the penalty's reference distribution); ship the `NotImplementedError` guard now and pick up the derivation in a separate PR.  Original test 5 lives there.
- **Bias-aware sandwich SEs for `θ_0`** -- not currently captured in a separate issue.  The MR1 sandwich is frequentist around `θ̂_pen`; if callers want coverage of `θ_0` they'd need either a bias correction or a fully-Bayesian treatment.  Filing on demand.
- **Composability documentation with `cue_ridge` / `cue_target_condition`** -- both knobs survive, but two ridges firing together is usually a sign one of them should be off.  Docstring + a paragraph in `weighting_info` will note the redundancy; basin-shifting consequences of stacking are #16's territory, not this issue's.

### Open design points (decide during MR1 review, not blockers)

- Whether the `PenaltyStrategy` Protocol should also accept a `tangent_basis` argument for non-Euclidean manifolds, or whether the penalty is always expressed in ambient-coordinate parameter space and the framework handles tangent-space conversion.  K-Aggregators' use case is Euclidean × symmetric-Γ; ambient-coords is sufficient for that.  A non-Euclidean-parameter penalty would want the tangent-basis form, but no current consumer needs it.
- Naming: `data_criterion_value` vs. `data_J` vs. `g_bar_W_g_bar_at_theta_hat`.  `data_criterion_value` is the least-bad of the three on this list.

Original issue body follows.

## Motivation

Some GMM problems need a parameter-space penalty on the estimator -- e.g.\ an L2 ridge on a slope sub-vector when the score has a near-null direction in that slope subspace.  In the K-Aggregators heterogeneous-\$\\theta\$ exp-link estimator we have exactly this situation: \`cond(D'WD)\` at the optimum is 4e9 (K=1) to 1e16 (K=2), the optimizer converges to economically-wild parameter values (per-product \$\\theta_j\$ across 21 orders of magnitude), and a small ridge on the slope coefficients restores interpretability.

There is no obvious place in the current API to express this.  The workaround we landed (K-Aggregators commit \`f3401d7\`) augments the moment vector with deterministic \"artificial moments\" \$g_{N+m}(\\theta) = \\sqrt{\\lambda} \\cdot c_m\$.  Under identity weighting this gives the right optimizer behaviour (\$\\lambda \\cdot \\sum_m c_m^2\$ added to the criterion).  But it has many problems that all share a root cause -- the penalty is the wrong kind of object to be a moment:

- Two-step / CUE / iterated weighting all attempt to invert \$\\hat\\Omega\$ that has a zero-variance block from the deterministic moments, so they NaN immediately.  Workaround: force \`IdentityWeighting\` whenever the penalty is on, which loses asymptotic efficiency on the data moments.
- \`compute_hessian_cond\`, \`k_statistic\`, \`tangent_covariance\` all see the augmented moment vector.  Degrees of freedom, sandwich SEs, K-statistic decomposition are all silently distorted by the deterministic block.  We dodged this only by not invoking those methods on the regularized pickles.
- \`criterion_value\` mixes data fit and penalty.  Recovering the data-only Hansen J for chi-square inference requires manual slicing + a reweight, which is error-prone (I got the N scaling wrong on first pass and double-counted observations).
- Every downstream caller has to know the penalty exists and route around it.  No upstream test coverage; it's purely application-level code.

In short, the artificial-moments trick is the wrong abstraction.  The penalty is a **cost addition**, not a moment.

## Proposed API

Add a \`penalty\` kwarg to \`GMM\`:

\`\`\`python
class GMM:
    def __init__(
        self,
        restriction: MomentRestriction,
        weighting: WeightingStrategy | None = None,
        optimizer: ... = None,
        initial_point: ... = None,
        cue_ridge: float = 0.0,
        cue_target_condition: float | None = None,
        penalty: Callable[[Any], float] | None = None,  # NEW
    ): ...

    def criterion(self, theta: Any) -> float:
        \"\"\"Optimizer cost: g'Wg + penalty.\"\"\"
        gWg = float(self._backend_dot(theta, self._weighting))
        pen = float(self._penalty(theta)) if self._penalty is not None else 0.0
        return gWg + pen

    def unpenalized_criterion(self, theta: Any) -> float:
        \"\"\"Hansen J at theta -- g'Wg only, no penalty term.\"\"\"
        return float(self._backend_dot(theta, self._weighting))
\`\`\`

The \`penalty\` callable takes the manifold point (matching the moment function's signature) and returns a scalar.  Differentiable via the same backend as the moments (JAX in our use case).  Composes naturally with all existing weighting strategies because it never touches \$\\hat\\Omega\$.

On \`GMMResult\`:

- \`criterion_value\` (existing) stays as the optimizer's stopping cost -- penalty included.
- New \`unpenalized_criterion_value\`: the chi-square-comparable Hansen J at \$\\hat\\theta\$, data moments only.
- \`compute_hessian_cond\`, \`k_statistic\`, \`tangent_covariance\` use the data Jacobian and \$\\hat\\Omega\$ unchanged for the moment-driven terms, and add the penalty's analytic Hessian for the information matrix entry.  The score-test decomposition in \`k_statistic\` should account for the penalty's contribution at \$\\theta_0\$ vs \$\\hat\\theta\$.

\`MomentRestriction\` does not change.  The penalty is its own object, parallel to the weighting and the optimizer.

## Why upstream rather than caller-side

The K-Aggregators use case is one application but the abstraction is general.  ManifoldGMM's existing facilities exactly cover what's needed for proper penalty handling:

- \`weighting_info\` already serializes per-stage diagnostics; can extend to record penalty config.
- \`tangent_covariance\` already builds the sandwich; needs the penalty Hessian as an additive correction.
- The pickle/refit machinery already handles per-stage weighting state; same pattern for the penalty.
- The CI tests already cover criterion/J consistency; a penalty test slot fits naturally.

Doing this at the caller level -- as we did in K-Aggregators -- silently breaks all four of those facilities.  Doing it upstream once makes parameter-space regularization a first-class feature available to all callers, with the same instrumentation guarantees as the rest of ManifoldGMM.

The deeper principle: we have full control over ManifoldGMM and should prefer changing how we ESTIMATE (which weighting, which optimizer, which CLI flag) over changing the ESTIMATOR itself (the moment vector definition).  The artificial-moments trick crossed that line; a native penalty hook moves the regularization back to the right side of the boundary.

## Test cases / acceptance criteria

1. **Closed-form ridge consistency**: a linear GMM problem (Euclidean manifold, linear moments) with an L2 penalty should recover the closed-form ridge regression estimate to machine precision, and report \`unpenalized_criterion_value\` matching the standard residual-sum-of-squares at the ridge \$\\hat\\beta\$.
2. **Penalty-free fallback**: \`penalty=None\` reproduces current behaviour bit-identically (criterion_value, k_statistic, etc.\ all unchanged).
3. **Cross-weighting**: same penalty under CUE, two-step, and iterated weighting gives consistent \`unpenalized_criterion_value\` at the optimum (data fit unchanged by weighting choice once both have converged to compatible \$\\hat\\theta\$ regions; weighting-stable \$\\hat\\theta\$ is a separate property).
4. **\`compute_hessian_cond\` distinguishes data vs penalty curvature**: a problem with near-singular \$D'WD\$ but a strictly convex penalty (\$\\nabla^2 \\text{penalty} \\succ 0\$) should have a well-conditioned full Hessian, surfacing through \`compute_hessian_cond\` correctly.  The data-only conditioning should remain reportable separately (e.g.\ via a new property or kwarg).
5. **\`k_statistic\` under penalty**: the score-test decomposition should be valid under a quadratic penalty; results should match Hansen's original derivation modulo the penalty's contribution to \$D'WD\$.  Tests against a small Monte Carlo where the design admits known-distribution K and S.
6. **Pickle round-trip**: a regularized \`GMMResult\` should pickle and unpickle without losing the penalty configuration, and \`unpenalized_criterion_value\` should be reproducible from the unpickled object.

## Adjacent fixes that drop out

- The hang from issue #18 (\`k_statistic\` infinite ridge-bump loop on near-singular \$D'\\Omega^{-1}D\$) is partly because the score test has no awareness of when the parameter is under-identified by the data moments.  Once a penalty hook exists, the curvature in the penalty-quadratic direction is well-defined even when \$D'\\Omega^{-1}D\$ alone has a near-null direction.  The penalized full Hessian inverts cleanly; the ridge-bump path doesn't need to fire.
- Issue #16 (CUE basin-shifting) becomes more diagnosable: with a known penalty contribution to the cost, the difference between \"CUE moved to a different basin\" and \"CUE is satisfying a different optimum\" can be isolated.

## Current workaround for callers

Pending this feature, K-Aggregators uses the artificial-moments trick (\`reports/exp_link_regularization_design.org\` in that repo).  When the upstream lands, that code reverts and the regularization moves into \`GMM(..., penalty=lambda theta: lambda_c * jnp.sum(theta[1][1:]**2))\`.

I'm filing this as a feature request rather than offering the PR myself -- the upstream contributors have the test infrastructure, the API design context, and the CI signal to do this cleanly.  Happy to provide the K-Aggregators repro and a sample real-world penalty (the het-exp Tikhonov) for the test suite.

**Comment by ligon** [comment]
MR1 PR opened as #22 against `main`.  Implements the scope-split section of this issue's body verbatim (PenaltyStrategy Protocol, `penalty=` plumbing, `data_criterion_value`, `compute_hessian_cond(data_only=)`, `tangent_covariance` with penalty Hessian in the bread, `k_statistic` NotImplementedError guard pointing at #21).  10 new tests; `make quick-check` green locally.

**Comment by ligon** [comment]
## Empirical validation -- K-Aggregators diagnostic log

For the record: a K-Aggregators session ran the #10 diagnostics (`compute_hessian_cond`, `optimizer_health`) against the exp-link runaway that motivated #19 and #20.  Full log lives in K-Aggregators (`logs/diag_runaway_20260522.jsonl`); the per-fit summary is below.

```json
{"pickle": "...mgmm_lf_K1.pickle", "het_theta_env": "exp", "J": 129.4209, "DoF": 22, "two_step": true,
 "mid": [41.5307, -4.0195, 7.9568, 15.1550, 13.2602, 15.5518],
 "hessian_cond": 4.062e9,
 "optimizer_health": {"n_outer_iters": 41, "inner_cap_hit_frac": 0.0, "n_inner_cap_hits": 0,
                      "tail_grad_slope": -0.6534, "tail_window": 20,
                      "inner_stop_counts": {"exceeded trust region": 13, "negative curvature": 2,
                                            "reached target residual-kappa (linear)": 21,
                                            "reached target residual-theta (superlinear)": 5}},
 "k_stat": null, "k_stat_skipped": true}

{"pickle": "...mgmm_lf_K2.pickle", "het_theta_env": "exp", "J": 23.6961, "DoF": 17, "two_step": true,
 "mid": [15.6566, -3.0092, 11.1511, -1.0143, -1.0409, -5.6019],
 "hessian_cond": 2.327e16,
 "optimizer_health": {"n_outer_iters": 37, "inner_cap_hit_frac": 0.0, "n_inner_cap_hits": 0,
                      "tail_grad_slope": -0.8069, "tail_window": 20,
                      "inner_stop_counts": {"negative curvature": 3, "exceeded trust region": 12,
                                            "reached target residual-kappa (linear)": 17,
                                            "reached target residual-theta (superlinear)": 5}},
 "k_stat": null, "k_stat_skipped": true}
```

### Three reads worth filing

1. **`compute_hessian_cond` is diagnostic of the pathology even when `optimizer_health` is benign.**  K=1 returns `4.06e9` (no rotation gauge to confound; unambiguous c-direction weakness across 9 orders of magnitude).  K=2 returns `2.33e16` -- machine-eps singular *after* the canonical tangent basis has modded out the symmetric-Gamma rotation gauge.  Both fits show zero inner-CG cap-hits, a steeply negative gradient slope (-0.65 / -0.81), and a healthy mix of inner stop reasons dominated by residual-target hits.  The runaway is a **post-convergence pathology** -- the optimizer settles into a stationary point of a near-flat criterion -- not a numerical-solver issue.  This is why #10's PR 2 stall warning correctly does not fire and #20's Hessian ridging would be the wrong remedy.
2. **`k_statistic` was already being routed around manually** (`k_stat_skipped: true` on both fits).  The K-Aggregators session knew not to invoke it on the regularized pickles.  #19 MR1's `NotImplementedError` guard formalizes what callers were already doing by hand.
3. **The fits are using the K-Aggregators artificial-moments workaround** (commit `f3401d7` over there).  Once PR #22 merges, that workaround can be retired and the Tikhonov-on-c sweep moves to `GMM(..., penalty=lambda c: lambda_c * jnp.sum(c[1:]**2))`.

### Consequence for issue triage

- **#10**: as scoped, the PR 2 stall warning doesn't generalize beyond MAX_INNER_ITER plateaus.  Worth a docstring caveat in `_maybe_warn_optimizer_health` ("targets cap-hit + flat-gradient signature; for post-convergence runaways, `compute_hessian_cond` is the diagnostic") at some point, but not blocking.
- **#20**: closed as not-planned -- see [this comment](https://github.com/ligon/ManifoldGMM/issues/20#issuecomment-4524188608).
- **#19 MR1 (PR #22)**: validated.  The diagnostic surface and the penalty hook together cover the pathology.
